src/extractor/extractor.cpp

```diff
@@ -808,5 +818,173 @@ void Extractor::WriteCompressedNodeBasedGraph(const std::string &path,
     }
 }
 
+
+struct EdgeInfo
+{
+  NodeID node;
+
+  uint32_t name_id;
+
+  // 0 - outgoing (forward), 1 - incoming (reverse), 2 - both outgoing and incoming
+  int direction;
+
+  ClassData road_class;
+
+  struct LessName
+  {
+    bool operator() (EdgeInfo const & e1, EdgeInfo const & e2) const
+    {
+      return e1.name_id < e2.name_id;
+    }
+  };
+};
+
+bool IsSegregated(std::vector<EdgeInfo> v1, std::vector<EdgeInfo> v2, EdgeInfo const & current)
+{
+  if (v1.size() < 2 || v2.size() < 2)
+    return false;
+
+  auto const preprocessFn = [] (std::vector<EdgeInfo> & v)
+  {
+    std::sort(v.begin(), v.end(), EdgeInfo::LessName());
+  };
+
+  preprocessFn(v1);
+  preprocessFn(v2);
+
+  // Internal edge with the name should be connected with any other neibour edge with the same name.
+  if (current.name_id != EMPTY_NAMEID)
+  {
+    auto const findNameFn = [&current] (std::vector<EdgeInfo> const & v)
+    {
+      return std::binary_search(v.begin(), v.end(), current, EdgeInfo::LessName());
```

Comment:
one thing I forgot to mention: name_ids are not necessarily unique for the same name:

`name: lincoln avenue` and `name: lincoln avenue, ref:I55` will result in two different name_ids, even though names remain the same. Adding a reference is a typical feature in OSM. You can check out `requiresNameAnnounced` in util for all name-related checks.